File: src/main/java/komu/blunt/types/patterns/Pattern.java
package komu.blunt.types.patterns;

import com.google.common.collect.ImmutableList;
import komu.blunt.objects.Symbol;
import komu.blunt.types.ConstructorNames;

import static komu.blunt.objects.Symbol.symbol;

public abstract class Pattern {

    @Override
    public abstract String toString();

    public static Pattern constructor(String name, Pattern... args) {
        //return new ConstructorPattern(name, ImmutableList.copyOf(args));
        return null;
    }
    
    public static Pattern constructor(String name, ImmutableList<Pattern> args) {
        //return new ConstructorPattern(name, args);
        return null;
    }

    public static Pattern variable(String name) {
        //return variable(symbol(name));
        return null;
    }

    public static Pattern variable(Symbol name) {
        //return new VariablePattern(name);
        return null;
    }

    public static Pattern literal(Object value) {
        //return new LiteralPattern(value);
        return null;
    }
    
    public static Pattern wildcard() {
        //return WildcardPattern.INSTANCE;
        return null;
    }

    public static Pattern tuple(ImmutableList<Pattern> args) {
        /*
        if (args.isEmpty())
            return constructor(ConstructorNames.UNIT);
        if (args.size() == 1)
            return args.get(0);
        else
            return constructor(ConstructorNames.tupleName(args.size()), args);
            */
        return null;
    }
}

File: src/main/kotlin/komu/blunt/types/patterns/Pattern.kt
package komu.blunt.types.patterns

import komu.blunt.objects.Symbol
import komu.blunt.types.ConstructorNames

sealed class Pattern {
    override abstract fun toString(): String

    companion object {


        fun variable(name: Symbol): Pattern = Variable(name)

        fun tuple(args: List<Pattern>): Pattern =
            when (args.size) {
                0    -> Constructor(ConstructorNames.UNIT)
                1    -> args.first()
                else -> Constructor(ConstructorNames.tupleName(args.size), args)
            }
    }

    class Constructor(val name: String, val args: List<Pattern>) : Pattern() {

        constructor(name: String, vararg args: Pattern): this(name, args.asList()) { }

        fun map(f: (Pattern) -> Pattern) = Constructor(name, args.map(f))

        override fun toString(): String =
                if (args.isEmpty())
                    name
                else
                    args.joinToString(" ", "($name", ")")

        override fun equals(other: Any?) = other is Constructor && name == other.name && args == other.args
        override fun hashCode() = name.hashCode() * 79 + args.hashCode()
    }

    object Wildcard : Pattern() {
        override fun toString() = "_"
    }

    class Variable(val variable: Symbol) : Pattern() {
        override fun toString() = variable.toString()
        override fun equals(other: Any?) = other is Variable && variable == other.variable
        override fun hashCode() = variable.hashCode()
    }

    class Literal(val value: Any) : Pattern() {
        override fun toString() = value.toString()
        override fun equals(other: Any?) = other is Literal && value == other.value
        override fun hashCode() = value.hashCode()
    }
}
